File: LeetCode/Medium/0433-minimum-genetic-mutation/0433-minimum-genetic-mutation.py
class Solution:
    def minMutation(self, startGene: str, endGene: str, bank: List[str]) -> int:
        ret = 1
        remained = set(bank)
        stack = [startGene]

        while stack:
            temp_stack = []
            while stack:
                curr = stack.pop()
                
                for i in range(8):
                    for v in "ACGT":
                        temp = curr[:i] + v + curr[i+1:]
                        
                        if temp in remained:
                            if temp == endGene:
                                return ret
                                
                            remained.remove(temp)
                            temp_stack.append(temp)
            ret += 1
            stack = temp_stack

        return -1

File: LeetCode/Medium/0433-minimum-genetic-mutation/0433-minimum-genetic-mutation.go
func minMutation(startGene string, endGene string, bank []string) int {
    ret := 1
    idx := 0
    stack := make([]string, 1)
    stack[0] = startGene
    used := make([]bool, len(bank))
    
    for idx < len(stack) {
        end := len(stack)
        
        for i := idx; i < end; i++ {
            curr := stack[i]
            
            for j := 0; j < len(bank); j++ {
                if used[j] { continue }
                
                count := 0
                
                for k := 0; k < 8; k++ {
                    if curr[k] != bank[j][k] { count++ }
                }
                
                if count == 1 {
                    if bank[j] == endGene { return ret }
                    used[j] = true
                    stack = append(stack, bank[j])
                }
            }
        }
        idx = end
        ret++
    }
    
    
    return -1
}